prov/shm/src/smr_progress.c

```diff
@@ -651,11 +651,20 @@ int smr_progress_unexp(struct smr_ep *ep,
 		entry->err = FI_EINVAL;
 	}
 
-	ret = smr_complete_rx(ep, entry->context, unexp_msg->cmd.msg.hdr.op,
-			unexp_msg->cmd.msg.hdr.op_flags | entry->flags,
+	comp_entry = *entry;
+	if (entry->flags & SMR_MULTI_RECV) {
+		smr_progress_multi_recv(ep,
+			unexp_msg->cmd.msg.hdr.op == ofi_op_tagged ?
+			&ep->trecv_queue : &ep->recv_queue, entry, &comp_entry,
+			total_len);
+		free_entry = (comp_entry.flags == entry->flags);
+	}
+
+	ret = smr_complete_rx(ep, comp_entry.context, unexp_msg->cmd.msg.hdr.op,
+			unexp_msg->cmd.msg.hdr.op_flags | comp_entry.flags,
 			total_len, entry->iov[0].iov_base,
-			unexp_msg->cmd.msg.hdr.addr, entry->tag,
-			unexp_msg->cmd.msg.hdr.data, entry->err);
+			unexp_msg->cmd.msg.hdr.addr, comp_entry.tag,
+			unexp_msg->cmd.msg.hdr.data, comp_entry.err);
```

Comment:
Similar comments as above.  comp_entry here can be replaced by entry, since the values are unchanged.  The only change is to the flags, so making a copy of entry isn't necessary.

I would change smr_progress_multi_recv() to take free_entry as input.  I would also change free_entry to a bool, since that's how it is used.